src/acl.ts

```diff
@@ -332,15 +340,96 @@ export function internal_combineAccessModes(
   );
 }
 
+/** @internal */
+export function internal_removeEmptyAclRules<
```

Comment:
Hmm crap. I don't think you can actually specify a Resource's access in a Resource other than its own ACL. However, I think there's a similar case that could be problematic:

>  The server would check if `/documents/.acl` exists, and then `/.acl`, until it finds some authorizations that contain `acl:default`.

In other words, having an empty `acl:default` rule would still prevent the server from traversing further up the tree. And what's worse, is that tracking which Rules we've modified doesn't help here: if there's just one rule setting the default for an agent, and we remove that agent from that rule, that doesn't mean that the developer suddenly wants the ACL from a higher-level Container to apply.

Reply:
> I don't think you can actually specify a Resource's access in a Resource other than its own ACL.
I would assume that would be true in the majority of cases, but if I'm playing the devil's advocate, why not ? The [ACL inheritence algorithm](https://github.com/solid/web-access-control-spec#acl-inheritance-algorithm) seems to say that if a Resource doesn't have an ACL document attached, you look up its parent's for "[a] set of Authorization statements to apply", which isn't contradictory with the parent containing a resource-scope statement about its child. In the worst case, you could imagine having one root ACL document that contains resource-scoped statements for all the resources inside a Pod. I'll ask the spec team to clarify.

Reply:
Logged here: https://github.com/solid/authorization-panel/issues/75

Reply:
> if there's just one rule setting the default for an agent, and we remove that agent from that rule, that doesn't mean that the developer suddenly wants the ACL from a higher-level Container to apply.

In any case, the presence of a `acl:default` rule, even applied to only one agent, stops the ACL inheritence algorithm, right ? So if we remove an agent `a` from a default rule `r` to a container `c`, the possible options I see are:
- Either we are giving `a` a different default access to `c`, so removing `r` once the new rule is created doesn't have unintended effects
- Or we are removing default access for `a` to `c`. In this case, deleting the empty `r` does have unintended effect. 

It seems that in order to avoid side-effects, we can only remove a default rule when another applies to the same resource (regardless of the agents involved). Is that correct ?

Reply:
> I would assume that would be true in the majority of cases, but if I'm playing the devil's advocate, why not ? The ACL inheritence algorithm seems to say that if a Resource doesn't have an ACL document attached, you look up its parent's for "[a] set of Authorization statements to apply", which isn't contradictory with the parent containing a resource-scope statement about its child. In the worst case, you could imagine having one root ACL document that contains resource-scoped statements for all the resources inside a Pod. I'll ask the spec team to clarify.

Yeah that's a good idea; it seems like that shouldn't be possible, but it does indeed leave room for that.

>  In any case, the presence of a acl:default rule, even applied to only one agent, stops the ACL inheritence algorithm, right ?

I'm not even sure if it needs to be applied to one agent...

> It seems that in order to avoid side-effects, we can only remove a default rule when another applies to the same resource (regardless of the agents involved). Is that correct ?

Hmm... Yeah, assuming that still just _one_ ACL can apply to a Resource, then we should be able to check whether a Rule to remove is the last one acting as the default for a Container. (And maybe even do the same for Resource Rules, to be sure?) But I'm not entirely confident when saying that :)

Reply:
Logged in https://github.com/solid/specification/issues/186

Reply:
OK, given [this comment by TimBL](https://github.com/solid/specification/issues/55#issuecomment-536092174) (pointed out by Aaron [here](https://github.com/solid/specification/issues/186#issuecomment-649781581)), whenever an ACL Resource _exists_ (it doesn't even have to contain _any_ Authorization Rules), that will prevent ACLs higher up in the tree from applying. Which means, I think, that we're safe here, no? :)

(Aaron also [mentioned](https://github.com/solid/authorization-panel/issues/75#issuecomment-649778752) that his implementation ignores all ACLs that define multiple target Resources. At first sight I thought that meant we could consider a simpler implementation: simply remove the Agent from the existing Rule(s), and create a new Rule that sets the desired Access Modes. But I just realised that that wouldn't actually work: you could still have a Rule that acts as both `acl:default` and `acl:accessTo`, so you'd still have to potentially copy over some of the original Rule iff it also acts as the type of Rule you're not changing. We could do an explicit check for that and recreate it only if necessary rather than copying it anyway and cleaning it up afterwards if it wasn't needed, but the current implementation feels safer to me. I'm logging this mostly for posterity :smiley:  .)

Reply:
If the presence of the ACL document, even empty, stops the inheritence recursion, then I think we're safe, because removing the rule cannot give extra permissions to the agent which has just been removed. As for the preventive copy/cleanup as opposed to a smarter solution, I feel like having something more systematic involves less edge cases, so I'm happy about the current implementation as well :)

Reply:
Btw, I just realised... If the fact that an ACL Resource exists is significant even if there are no Quads in there, that does mean that @pmcb55's idea of ACLs being available at the same URL to enable atomic updates is not going to be possible, is it?

Reply:
And here's another confirmation of one-resource-per-ACL (because associated ACL Resources need to be deleted whenever the applicable Resource gets deleted): https://github.com/solid/authorization-panel/issues/75#issuecomment-651629554

Reply:
And just to log it in a public place, now that the comment is available again (GH had some issues yesterday), here is a snippet of our discussion about atomic updates.

I'd say it depends whether a Named Graph may be declared even if it contains no triples. The tricky part is that depending on the RDF syntax, the answer is not always obvious:
- In TriG,
```
<http://example.org/my-resource/my-acl-graph>
{
# No triples, but the graph is there
}
```
does make sense
- In NQuads,
```
??? ??? ??? <http://example.org/my-resource/my-acl-graph>
```
is a bigger issue.
In Turtle, that doesn't have the notion of Named Graph, I don't see how that would work. A possible approach would be to extend the metadata model so that when looking up a resource with a `Response-Type` set to an RDF serialization that supports named graphs (e.g. TriG or NQuads), you get the resource and its metadata as named graphs in a single request (and similarly you can perform a single write), and with RDF serializations that don't support that you must cannot have atomic interaction with the resource+metadata. That would both preserve backward compatibility and support the feature, but it's quite a long shot on top of a still-not-specified feature.

Vinnl: will the server behave differently when there are empty Named Graphs? And if it will, that will be problematic if there is no way to encode that in one of the mandated serialisation formats.

NSeydoux: I think that (even more since Turtle is suppoted now) the server cannot expect ACLs updates to be always atomic. But it would be interesting to consider if the metadata resource model allows to put everything in one body, basically having the server concatenate several resources located at different IRIs into one response (instead of providing Link headers to these resources), and to authorize the client to likewise upload at once the resource and its associated resources

That may turn out to be a disaster because it creates huge consistency problems and all, but that could be one approach